# Sampling

## Nivell 1

### Exercise 1 - sample random simple and sample sistematic

Dataset about earnings in esports:
- Kaggle: https://www.kaggle.com/jackdaoud/esports-earnings-for-players-teams-by-game 
- Working with the amount of earnings from each player to represent the statistics about the highest earnings. 
- Dataset contains 1000 total observations from the top 10 games on eSports.

In [23]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import random
import statistics
%matplotlib inline
from sklearn.model_selection import train_test_split
from random import randint

In [24]:
esports_df = pd.read_csv('highest_earning_players.csv')

In [25]:
esports_df.head(5)

PlayerId NameFirst   NameLast CurrentHandle CountryCode  TotalUSDPrize  \
0      3883     Peter  Rasmussen       dupreeh          dk     1822989.41   
1      3679   Andreas   Højsleth         Xyp9x          dk     1799288.57   
2      3885   Nicolai     Reedtz        dev1ce          dk     1787489.88   
3      3672     Lukas  Rossander        gla1ve          dk     1652350.75   
4     17800      Emil       Reif        Magisk          dk     1416448.64   

                               Game                 Genre  
0  Counter-Strike: Global Offensive  First-Person Shooter  
1  Counter-Strike: Global Offensive  First-Person Shooter  
2  Counter-Strike: Global Offensive  First-Person Shooter  
3  Counter-Strike: Global Offensive  First-Person Shooter  
4  Counter-Strike: Global Offensive  First-Person Shooter

In [26]:
esports_df.describe()

PlayerId  TotalUSDPrize
count   1000.000000   1.000000e+03
mean   27793.587000   3.977932e+05
std    22170.225194   6.908492e+05
min     1000.000000   2.417167e+04
25%     5374.500000   8.378962e+04
50%    23502.000000   1.683284e+05
75%    48127.250000   3.937352e+05
max    83085.000000   6.952597e+06

#### Simple random sample

In [27]:
simple_esports_df = esports_df.sample(n = 100)
simple_esports_df.head(5)

PlayerId NameFirst  NameLast CurrentHandle CountryCode  TotalUSDPrize  \
201      6466  Ho Seong       Lee          Duke          kr      954620.62   
753     66881      Luke     Newey        Luke12          au       79049.82   
968     33858    Thomas    Zimmer      Sintolol          de       76346.78   
101      3822     Jesse  Vainikka         JerAx          fi     6470000.02   
20       3290    Nathan   Schmitt           NBK          fr      805444.66   

                                 Game                            Genre  
201                 League of Legends  Multiplayer Online Battle Arena  
753                              PUBG                    Battle Royale  
968                       Hearthstone            Collectible Card Game  
101                            Dota 2  Multiplayer Online Battle Arena  
20   Counter-Strike: Global Offensive             First-Person Shooter

In [28]:
simple_esports_df.info()#I want to check if I really have 100 values

<class 'pandas.core.frame.DataFrame'>
Int64Index: 100 entries, 201 to 283
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PlayerId       100 non-null    int64  
 1   NameFirst      100 non-null    object 
 2   NameLast       100 non-null    object 
 3   CurrentHandle  100 non-null    object 
 4   CountryCode    100 non-null    object 
 5   TotalUSDPrize  100 non-null    float64
 6   Game           100 non-null    object 
 7   Genre          100 non-null    object 
dtypes: float64(1), int64(1), object(6)
memory usage: 7.0+ KB


In [29]:
simple_esports_df.describe()

PlayerId  TotalUSDPrize
count    100.000000   1.000000e+02
mean   32068.860000   3.513494e+05
std    23037.627072   7.081882e+05
min     1011.000000   3.617167e+04
25%    10669.500000   8.590563e+04
50%    30401.500000   1.777067e+05
75%    53840.250000   3.394026e+05
max    80697.000000   6.470000e+06

Interpretation: The arithmetic mean is 400498.9 USD prize by player. For a very simple statistics it might be ok but we need to get average rate and correctly weight each data point. Let's use the harmonic mean. It will help to lower the weight of the outliers without removing them.

In [30]:
har_mean1 = statistics.harmonic_mean(simple_esports_df['TotalUSDPrize'])
print('The harmonic mean is',round(har_mean1,2),'USD prize by player.')

The harmonic mean is 129909.23 USD prize by player.


#### Systematic sample

In [31]:
#random step number
step_num = 10

def randomNum():
    return np.random.randint(1, 50)
num = randomNum()

In [32]:
#[start:stop:step]
systematic_esports_df = esports_df[num:len(esports_df):step_num]
systematic_esports_df.head(5)

PlayerId NameFirst      NameLast CurrentHandle CountryCode  TotalUSDPrize  \
7      12169  Fernando     Alvarenga           fer          br     1063038.92   
17      5000    Freddy     Johansson         KRiMZ          se      867823.34   
27      1511    Daniil      Teslenko          Zeus          ua      689568.70   
37      1488  Jarosław  Jarząbkowski   pashaBiceps          pl      577382.35   
47     12859     Jonas      Olofsson         Lekr0          se      464499.26   

                                Game                 Genre  
7   Counter-Strike: Global Offensive  First-Person Shooter  
17  Counter-Strike: Global Offensive  First-Person Shooter  
27  Counter-Strike: Global Offensive  First-Person Shooter  
37  Counter-Strike: Global Offensive  First-Person Shooter  
47  Counter-Strike: Global Offensive  First-Person Shooter

In [33]:
systematic_esports_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 7 to 997
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PlayerId       100 non-null    int64  
 1   NameFirst      100 non-null    object 
 2   NameLast       100 non-null    object 
 3   CurrentHandle  100 non-null    object 
 4   CountryCode    100 non-null    object 
 5   TotalUSDPrize  100 non-null    float64
 6   Game           100 non-null    object 
 7   Genre          100 non-null    object 
dtypes: float64(1), int64(1), object(6)
memory usage: 6.4+ KB


I have only 96 entries so I would like to add the rest to get 100 entries with the same step.

In [34]:
len(systematic_esports_df)

100

In [35]:
addNum = (100 - len(systematic_esports_df))
addNum

0

In [36]:
x = step_num - addNum
var1 = esports_df.iloc[x]
var2 = esports_df.iloc[x+step_num]
var3 = esports_df.iloc[x+(2*step_num)]

In [37]:
systematic_esports_df = systematic_esports_df.append(var1, ignore_index = True)
systematic_esports_df = systematic_esports_df.append(var2, ignore_index = True)
systematic_esports_df = systematic_esports_df.append(var3, ignore_index = True)

In [38]:
systematic_esports_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103 entries, 0 to 102
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PlayerId       103 non-null    int64  
 1   NameFirst      103 non-null    object 
 2   NameLast       103 non-null    object 
 3   CurrentHandle  103 non-null    object 
 4   CountryCode    103 non-null    object 
 5   TotalUSDPrize  103 non-null    float64
 6   Game           103 non-null    object 
 7   Genre          103 non-null    object 
dtypes: float64(1), int64(1), object(6)
memory usage: 6.6+ KB


Now I have 100 entries and I can compare harmonic mean of two methods of sampling.

In [39]:
har_mean2 = statistics.harmonic_mean(systematic_esports_df['TotalUSDPrize'])
print('The harmonic mean is',round(har_mean2,2),'USD prize by player.')

The harmonic mean is 122484.65 USD prize by player.


## Nivell 2

### Exercise 2 - Stratified Random Sample and SMOTE

#### Stratified random sample - by algorithm

For Stratified Random Sample I will divide the player in groups by game, there are 10 games in the dataset, so I can pick 10 players form each(10%), than I will get 100 players in total. Use this sampling method when we have some minority group.

In [40]:
#how many inputs there are for each game
esports_df['Game'].value_counts()

PUBG                                100
Fortnite                            100
Hearthstone                         100
Starcraft II                        100
Counter-Strike: Global Offensive    100
Arena of Valor                      100
League of Legends                   100
Dota 2                              100
Heroes of the Storm                 100
Overwatch                           100
Name: Game, dtype: int64

In [41]:
#games
esports_df['Game'].unique()

array(['Counter-Strike: Global Offensive', 'Dota 2', 'League of Legends',
       'Fortnite', 'Overwatch', 'Starcraft II', 'Heroes of the Storm',
       'PUBG', 'Arena of Valor', 'Hearthstone'], dtype=object)

In [42]:
#stratified split - final 100 values
Xtrain, Xtest = train_test_split(esports_df, test_size=0.10, stratify=esports_df[['Game']])

In [43]:
#sample contains which games?
Xtest['Game'].unique()

array(['Overwatch', 'Starcraft II', 'Fortnite',
       'Counter-Strike: Global Offensive', 'Heroes of the Storm',
       'Dota 2', 'PUBG', 'Hearthstone', 'League of Legends',
       'Arena of Valor'], dtype=object)

In [44]:
Xtest['Game'].describe()

count           100
unique           10
top       Overwatch
freq             10
Name: Game, dtype: object

Sample is stratified - contains 80 values for each 10 games.

#### Stratified random sample - manually collected values

In [45]:
sorted_esports_df = esports_df. sort_values(by='Game')

In [46]:
sorted_esports_df.head(5)

PlayerId NameFirst       NameLast CurrentHandle CountryCode  \
801     60565    Guixin            Guo        NuoYan          cn   
873     75020       Aum  Chatchanapong         Happy          th   
872     54012   Denchai     Eampaojeen        JameCo          th   
871     53974         -              -       Sirenia          tw   
870     53973         -              -         Liang          tw   

     TotalUSDPrize            Game                            Genre  
801      416162.04  Arena of Valor  Multiplayer Online Battle Arena  
873       40376.87  Arena of Valor  Multiplayer Online Battle Arena  
872       41144.36  Arena of Valor  Multiplayer Online Battle Arena  
871       42482.43  Arena of Valor  Multiplayer Online Battle Arena  
870       42482.43  Arena of Valor  Multiplayer Online Battle Arena

In [47]:
sorted_esports_df['Game'].unique()

array(['Arena of Valor', 'Counter-Strike: Global Offensive', 'Dota 2',
       'Fortnite', 'Hearthstone', 'Heroes of the Storm',
       'League of Legends', 'Overwatch', 'PUBG', 'Starcraft II'],
      dtype=object)

In [48]:
group1 = sorted_esports_df[sorted_esports_df['Game'] == 'Arena of Valor']
group1.info()#to check if I get 100 entruies for this group, than I will apply the code for all 10 groups

<class 'pandas.core.frame.DataFrame'>
Int64Index: 100 entries, 801 to 802
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PlayerId       100 non-null    int64  
 1   NameFirst      100 non-null    object 
 2   NameLast       100 non-null    object 
 3   CurrentHandle  100 non-null    object 
 4   CountryCode    100 non-null    object 
 5   TotalUSDPrize  100 non-null    float64
 6   Game           100 non-null    object 
 7   Genre          100 non-null    object 
dtypes: float64(1), int64(1), object(6)
memory usage: 7.0+ KB


In [49]:
group2=sorted_esports_df[sorted_esports_df['Game'] == 'Counter-Strike: Global Offensive'] 
group3=sorted_esports_df[sorted_esports_df['Game'] == 'Dota 2'] 
group4=sorted_esports_df[sorted_esports_df['Game'] == 'Fortnite'] 
group5=sorted_esports_df[sorted_esports_df['Game'] == 'Hearthstone'] 
group6=sorted_esports_df[sorted_esports_df['Game'] == 'Heroes of the Storm'] 
group7=sorted_esports_df[sorted_esports_df['Game'] == 'League of Legends'] 
group8=sorted_esports_df[sorted_esports_df['Game'] == 'Overwatch'] 
group9=sorted_esports_df[sorted_esports_df['Game'] == 'PUBG'] 
group10=sorted_esports_df[sorted_esports_df['Game'] == 'Starcraft II']

In [50]:
#getting 10 entries from each group
esports_df1 = group1.sample(n = 10)
esports_df2 = group2.sample(n = 10)
esports_df3 = group3.sample(n = 10)
esports_df4 = group4.sample(n = 10)
esports_df5 = group5.sample(n = 10)
esports_df6 = group6.sample(n = 10)
esports_df7 = group7.sample(n = 10)
esports_df8 = group8.sample(n = 10)
esports_df9 = group9.sample(n = 10)
esports_df10 = group10.sample(n = 10)

In [51]:
stratified_esports_df = esports_df1.append((esports_df2,
                                            esports_df3,
                                            esports_df4,
                                            esports_df5,
                                            esports_df6,
                                            esports_df7,
                                            esports_df8,
                                            esports_df9,
                                            esports_df10),ignore_index = True)

In [52]:
stratified_esports_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PlayerId       100 non-null    int64  
 1   NameFirst      100 non-null    object 
 2   NameLast       100 non-null    object 
 3   CurrentHandle  100 non-null    object 
 4   CountryCode    100 non-null    object 
 5   TotalUSDPrize  100 non-null    float64
 6   Game           100 non-null    object 
 7   Genre          100 non-null    object 
dtypes: float64(1), int64(1), object(6)
memory usage: 6.4+ KB


In [53]:
har_mean3 = statistics.harmonic_mean(stratified_esports_df['TotalUSDPrize'])
print('The harmonic mean is',round(har_mean3,2),'USD prize by player.')

The harmonic mean is 123987.91 USD prize by player.


Interpretation: While applying stratified sampling method I've got similar value to simple sampling method.

#### SMOTE

In [54]:
import imblearn
from imblearn.over_sampling import SMOTE

In [55]:
esports_df['CountryCode'].unique()

array(['dk', 'us', 'br', 'ca', 'se', 'fr', 'sk', 'no', 'ba', 'ru', 'ua',
       'pl', 'nl', 'ee', 'fi', 'kz', 'bg', 'cz', 'tr', 'be', 'es', 'de',
       'gb', 'cn', 'uz', 'au', 'jo', 'lb', 'pk', 'il', 'ro', 'my', 'sg',
       'mo', 'mk', 'ph', 'kr', 'by', 'tw', 'hr', 'hk', 'si', 'at', 'ar',
       'ch', 'lv', 'lt', 'jp', 'mx', 'it', 'ie', 'gi', 'cu', 'vn', 'th',
       'gr'], dtype=object)

In [56]:
esports_df.groupby(['CountryCode']).size()#how many player I have by country? I will pick two countries for SMOTE

CountryCode
ar      3
at      3
au      5
ba      1
be      5
bg      6
br     10
by      2
ca     37
ch      3
cn    168
cu      1
cz      3
de     30
dk     36
ee      2
es      6
fi     16
fr     37
gb     16
gi      1
gr      1
hk      5
hr      2
ie      1
il      2
it      2
jo      2
jp      4
kr    251
kz      6
lb      1
lt      2
lv      1
mk      1
mo      1
mx      1
my      6
nl     10
no     10
ph      7
pk      1
pl     14
ro      2
ru     22
se     45
sg      1
si      4
sk      3
th     29
tr      4
tw     34
ua     12
us    114
uz      1
vn      7
dtype: int64

I will take players from South Korea - KR and China - CN because these countries has the highest number of players.

In [57]:
groupKR=esports_df[esports_df['CountryCode'] == 'kr']
groupCN=esports_df[esports_df['CountryCode'] == 'cn']

In [58]:
smote_esports_df = groupKR.append(groupCN,ignore_index = True)

In [59]:
smote_esports_df.head(5)

PlayerId   NameFirst NameLast CurrentHandle CountryCode  TotalUSDPrize  \
0      4885    Yong Min      Kim         Febby          kr      602221.88   
1      4888     Sunyeob      Kim            QO          kr      591322.79   
2      4883        Noah      Pyo            MP          kr      591176.77   
3      3832  Sang Hyeok      Lee         Faker          kr     1257615.87   
4      6466    Ho Seong      Lee          Duke          kr      954620.62   

                Game                            Genre  
0             Dota 2  Multiplayer Online Battle Arena  
1             Dota 2  Multiplayer Online Battle Arena  
2             Dota 2  Multiplayer Online Battle Arena  
3  League of Legends  Multiplayer Online Battle Arena  
4  League of Legends  Multiplayer Online Battle Arena

In [60]:
mod_smote_esports_df = smote_esports_df[['CountryCode','TotalUSDPrize']]
mod_smote_esports_df.head()

CountryCode  TotalUSDPrize
0          kr      602221.88
1          kr      591322.79
2          kr      591176.77
3          kr     1257615.87
4          kr      954620.62

In [61]:
pd.options.mode.chained_assignment = None
mod_smote_esports_df['CountryCode'] = np.where((mod_smote_esports_df['CountryCode'] == 'kr'), 1, 0)

In [62]:
mod_smote_esports_df

CountryCode  TotalUSDPrize
0              1      602221.88
1              1      591322.79
2              1      591176.77
3              1     1257615.87
4              1      954620.62
..           ...            ...
414            0       69487.73
415            0       68162.50
416            0       59309.90
417            0       54320.72
418            0       51231.70

[419 rows x 2 columns]

In [63]:
mod_smote_esports_df['CountryCode'].value_counts()

1    251
0    168
Name: CountryCode, dtype: int64

That means that I have 251 player from South Korea and 168 player from China. With SMOTE I will increase the values of China player to 251.

In [64]:
smote = SMOTE(sampling_strategy = 'minority')
X_sm, y_sm = smote.fit_resample(mod_smote_esports_df, mod_smote_esports_df['CountryCode'])

In [65]:
X_sm['CountryCode'].value_counts()

0    251
1    251
Name: CountryCode, dtype: int64

In [66]:
har_mean4 = statistics.harmonic_mean(X_sm['TotalUSDPrize'])
print('The harmonic mean is',round(har_mean4,2),'USD prize by player in China and South Korea.')

The harmonic mean is 117318.15 USD prize by player in China and South Korea.


## Nivell 3

### Exercise 3 - Reservoir samplig

Process:
- Tutorial: https://towardsdatascience.com/the-5-sampling-algorithms-every-data-scientist-need-to-know-43c7bc11d17c 
- Use for streaming, real time sampling.

In [67]:
def generator(max):
    number = 1
    while number < max:
        number += 1
        yield number
        
stream = esports_df['TotalUSDPrize']

k=100 #number of samples
reservoir = []#list of values
for i, element in enumerate(stream):
    if i+1<= k:
        reservoir.append(element)
    else:
        probability = k/(i+1)
        if random.random() < probability:
            # select item in stream and remove one of the k items already selected
            reservoir[random.choice(range(0,k))] = element
print(reservoir)

[883148.24, 136820.0, 111540.17, 1652350.75, 1416448.64, 1087340.0, 89346.12, 348691.73, 171767.15, 127688.41, 67687.33, 964635.39, 166127.4, 75756.09, 210416.67, 164821.18, 81932.0, 157289.9, 1213576.66, 65922.73, 111871.29, 167697.53, 337832.12, 388919.99, 83622.52, 159402.36, 506099.73, 122693.28, 86881.37, 75380.4, 61350.0, 94073.0, 63634.48, 31489.38, 221752.32, 54290.66, 143908.34, 98723.19, 153763.01, 122860.21, 913084.7, 105541.2, 488379.08, 141662.54, 47602.61, 60502.6, 85243.75, 51203.24, 329318.02, 1348804.67, 1533698.67, 2002159.51, 97741.11, 119560.82, 1021000.0, 5554297.41, 133558.02, 64373.86, 36171.67, 58160.23, 228500.0, 265283.06, 91850.98, 349557.33, 90485.05, 538926.12, 182066.13, 128681.89, 265601.98, 45995.54, 75125.32, 1923774.23, 307129.97, 184811.62, 61027.02, 189482.52, 134300.51, 794627.39, 91158.55, 121255.92, 591176.77, 229850.77, 65922.73, 66730.8, 269446.78, 162017.51, 64763.4, 66517.5, 260721.09, 76346.78, 314548.25, 591271.39, 2244012.27, 86277.6, 85032

In [68]:
len(reservoir)

100

In [69]:
a = min(reservoir)
b = max(reservoir)
print(a, b)

31489.38 5554297.41


In [70]:
har_mean5 = statistics.harmonic_mean(reservoir)
print('The harmonic mean is',round(har_mean5,2),'USD prize by player.')

The harmonic mean is 123246.36 USD prize by player.


In [71]:
print('Average prize by simple sampling is', round(har_mean1,2),'USD.')
print('Average prize by sistemating sampling is', round(har_mean2,2),'USD.')
print('Average prize by stratified sampling is', round(har_mean3,2),'USD.')
print('Average prize by SMOTE sampling is', round(har_mean4,2),'USD.')
print('Average prize by reservoir sampling is', round(har_mean5,2),'USD.')

Average prize by simple sampling is 129909.23 USD.
Average prize by sistemating sampling is 122484.65 USD.
Average prize by stratified sampling is 123987.91 USD.
Average prize by SMOTE sampling is 117318.15 USD.
Average prize by reservoir sampling is 123246.36 USD.
